### ISSUE TITLE
`declaration-colon-newline-after` triggers error for Sass maps and when aligning values (multiple preceding spaces)

### ISSUE THREAD

**Issue by deepfriedmind** [stylelint/stylelint#767]
I'm extending `stylelint-config-standard` which sets `"declaration-colon-newline-after": "always-multi-line"`. As this is something I agree with when it comes to properties/values, I want to keep the setting.

However, I've experienced 2 issues with this which I would like to resolve:

### Issue 1

The rule is triggered for multi-line Sass maps (SCSS syntax), e.g.:

```scss
$map: (
    key1: value1,
    key2: value2,
    key3: value3
);
```

### Issue 2

I like to align values for readability, e.g.:

```css
@font-face {
    font-family:  'Some font';
    font-variant: normal;
    font-weight:  normal;
    src:          url('../fonts/filename.woff2') format('woff2'),
                  url('../fonts/filename.woff') format('woff');
}
```

However, this triggers the rule because of the multiple spaces before the last `url()`.

Are these issues something that could be fixed or worked around somehow without disabling the rule?

**Comment by jeddy3** [comment]
Issue 1: If we can reliably distinguish SCSS maps from standard declarations then I think we could throw a conditional into the rule to ignore SCSS maps.

Am I right in thinking that we could identify a SCSS map as a declaration whom's parent is the root node, whom's `prop` starts with a `$`, and whom's `value` begins with a `(` and ends with `)`. (Ref: see this AST - https://astexplorer.net/#/u23yB7arjd). Can we think of anything in standard CSS that might have this signature?

@deepfriedmind In the meantime, wrapping your CSS maps in [*disabling comments*](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/configuration.md#turning-rules-on-and-off) will allow you to keep the rule on for standard declarations.

Issue 2: I think the rule is behaving as expected. `"declaration-colon-newline-after": "always-multi-line"` enforces this style:

```css
.selector {
  display: block;
  box-shadow:
    1px 1px 1px #000,
    2px 2px 1px 1px #ccc inset;
}
```

Note the newline after the colon within the multiline `box-shadow` declaration.

So if you want:

```css
@font-face {
  font-family: 'Some font';
  font-variant: normal;
  font-weight: normal;
  src: url('../fonts/filename.woff2') format('woff2'),
       url('../fonts/filename.woff') format('woff');
}
```

Use `"declaration-colon-space-after": "always"` and don't use the `"declaration-colon-newline-after"` rule. But if you want to align your values you could build on the recommendation in https://github.com/stylelint/stylelint/issues/769 and also check that the values *within value lists* also align :)

**Comment by deepfriedmind** [comment]
@jeddy3 Thanks for the in-depth reply. When it comes to Sass maps, you probably can't assume the map's direct parent is the root node, e.g. in this example: https://astexplorer.net/#/IxdCNwMPVR/1 – the other assumptions should be solid identifiers though and I can't think of anything from vanilla CSS which would overlap with this.

I'll look in to the plugin API as soon as I get some free time.

**Comment by jeddy3** [comment]
> When it comes to Sass maps, you probably can't assume the map's direct parent is the root node,

Good example, thanks :)

> the other assumptions should be solid identifiers though and I can't think of anything from vanilla CSS which would overlap with this.

Yep, `$` seems pretty unique to processors. I think parentheses are only used for functions in CSS and I think they always start with a name e.g. `prop: calc(1 * 2);`.  

@davidtheclark Solid enough to add this exclusion?

**Comment by davidtheclark** [comment]
I think that `$` is not a valid character for CSS identifiers, so it only has purpose with processors.

Looks like the AST generated by that map should be a declaration with a `$map` property name. So why not just add an option to ignore `$map` declarations? Seems very simple to fix this use-case.

**Comment by deepfriedmind** [comment]
@davidtheclark `$map` is just a variable name. The `prop` value could be `$anything`.

**Comment by davidtheclark** [comment]
Oh I see. My mistake. So we'd have to rely on it being a `$` variable starting and ending with parentheses. I might add a condition for it containing at least one colon inside the value.

**Comment by deepfriedmind** [comment]
That sounds great!

**Comment by jeddy3** [comment]
> I might add a condition for it containing at least one colon inside the value.

Sgtm.

You've added the "type: option" label. Do you mean this to be an option, or will SCSS maps be ignore by default ([like we do](https://github.com/stylelint/stylelint/blob/84bad43407f611bffd0d8768973105e9389b80a2/src/rules/color-no-invalid-hex/__tests__/index.js#L21) with things like variable interpolation)?

**Comment by davidtheclark** [comment]
Oh. Man I'm not very sharp this morning :) Yeah, I guess that I also can't
think of any danger in ignoring them by default, so this should be chanced
to an enhancement.

**Comment by jeddy3** [comment]
> Man I'm not very sharp this morning 

I was felt exactly the same way when I did that borked release yesterday morning :)

**Comment by ntwb** [comment]
FWIW I borked 2 release a couple of days ago :disappointed:

**Comment by jeddy3** [comment]
@deepfriedmind FYI, released in `4.4.0`.

**Comment by deepfriedmind** [comment]
Great, thanks!

**Comment by thasmo** [comment]
This doesn't seem to work for me (or I'm doing something wrong).

This is my `.stylelintrc` file:
```yml
extends: stylelint-config-standard
rules:
  indentation: tab
````
This is my map:
```css
$base-layers: (
	'footer',
	'header',
);
````
stylelint prints this error:
```
 1:13  ×  Expected newline after ":" with a multi-line value   declaration-colon-newline-after
```

**Comment by jeddy3** [comment]
Have you [correctly set](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/css-processors.md#parsing-non-standard-syntax) the syntax option, and are you running the latest version of stylelint?

**Comment by thasmo** [comment]
@jeddy3, yep using `gulp-stylelint` which installes version `6.1.1` and I've set the syntax to `scss`.

**Comment by jeddy3** [comment]
Thanks for letting me know.

I can confirm that the following construct does through an warning:

```
$base-layers: (
    'footer',
    'header',
);
```

I couldn't find anything in the sass docs what that construct is though. Isn't a map a series of key value pairs e.g. `$map: (key: value, key2: value2)`? That's what we're currently checking for. If the above construct is also a valid map, we can add it to definition. Otherwise, let me know what it's called and we can create a function that ignores it.

**Comment by thasmo** [comment]
@jeddy3 It's a [list](http://hugogiraudel.com/2013/07/15/understanding-sass-lists/).

**Comment by jeddy3** [comment]
Sweet. I just spotted it myself. I'll create a new issue for it.
